src/bmffimage.cpp

```diff
@@ -0,0 +1,502 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, 5th Floor, Boston, MA 02110-1301 USA.
+ */
+
+// *****************************************************************************
+
+#define EXIV2_DEBUG_MESSAGES
+
+// included header files
+#include "bmffimage.hpp"
+
+#include "basicio.hpp"
+#include "config.h"
+#include "error.hpp"
+#include "futils.hpp"
+#include "image.hpp"
+#include "image_int.hpp"
+#include "safe_op.hpp"
+#include "tiffimage.hpp"
+#include "tiffimage_int.hpp"
+#include "types.hpp"
+#include "unused.h"
+
+// + standard includes
+#include <cassert>
+#include <cstdio>
+#include <cstring>
+#include <iostream>
+#include <string>
+
+struct BmffBoxHeader
+{
+    uint32_t length;
+    uint32_t type;
+};
+
+#define TAG_ftyp 0x66747970 /**< "ftyp" File type box */
+#define TAG_avif 0x61766966 /**< "avif" AVIF */
+#define TAG_heic 0x68656963 /**< "heic" HEIC */
+#define TAG_heif 0x68656966 /**< "heif" HEIF */
+#define TAG_crx 0x63727820  /**< "crx " Canon CR3 */
+#define TAG_moov 0x6d6f6f76 /**< "moov" Movie */
+#define TAG_meta 0x6d657461 /**< "meta" Metadata */
+#define TAG_mdat 0x6d646174 /**< "mdat" Media data */
+#define TAG_uuid 0x75756964 /**< "uuid" UUID */
+#define TAG_dinf 0x64696e66 /**< "dinf" Data information */
+#define TAG_iprp 0x69707270 /**< "iprp" Item properties */
+#define TAG_ipco 0x6970636f /**< "ipco" Item property container */
+#define TAG_iinf 0x69696e66 /**< "iinf" Item info */
+#define TAG_iloc 0x696c6f63 /**< "iloc" Item location */
+#define TAG_ispe 0x69737065 /**< "ispe" Image spatial extents */
+#define TAG_infe 0x696e6665 /**< "infe" Item Info Extention */
+#define TAG_ipma 0x69706d61 /**< "ipma" Item Property Association */
+#define TAG_cmt1 0x434d5431 /**< ifd0Id */
+#define TAG_cmt2 0x434D5432 /**< exifID */
+#define TAG_cmt3 0x434D5433 /**< canonID */
+#define TAG_cmt4 0x434D5434 /**< gpsID */
+
+// *****************************************************************************
+// class member definitions
+namespace Exiv2
+{
+    static bool enabled = false;
+
+    EXIV2API bool enableBMFF(bool enable)
+    {
+#ifdef EXV_ENABLE_BMFF
+        enabled = enable;
+        return true;
+#endif                   // EXV_ENABLE_BMFF
+        enable = false;  // unused
+        return enable;
+    }
+
+    class Iloc
+    {
+    public:
+        Iloc(uint32_t ID = 0, uint32_t start = 0, uint32_t length = 0) : ID_(ID), start_(start), length_(length){};
+        virtual ~Iloc(){};
+
+        uint32_t ID_;
+        uint32_t start_;
+        uint32_t length_;
+
+        std::string toString() const
+        {
+            return Internal::stringFormat("ID = %u from,length = %u,%u", ID_, start_, length_);
+        }
+    };  // class Iloc
```

Comment:
Since everything is `public` anyway, how about turning this into a `struct` and then we can remove the `public` keyword. 
I also don't think we need the virtual destructor, do we? Would anyone really try and inherit from this class?
```suggestion
    struct Iloc
    {
        Iloc(uint32_t ID = 0, uint32_t start = 0, uint32_t length = 0) : ID_(ID), start_(start), length_(length){};

        uint32_t ID_;
        uint32_t start_;
        uint32_t length_;

        std::string toString() const
        {
            return Internal::stringFormat("ID = %u from,length = %u,%u", ID_, start_, length_);
        }
    };
```

Reply:
Sure.  Use a struct if you prefer.  I think most of the structs in the code-base are really good old "C" structs and don't have methods, constructors and destructors.  However some do.

Reply:
We don't need the virtual destructor until we need the virtual destructor because we've derived a new class from this.  So I think it's good practice to always declare virtual destructors.

There's no reason to be concerned about performance or memory demands from this code.